### DISCUSSION TITLE
Is there a way to link an external script in a Hyperdiv app?

### DISCUSSION THREAD

**Discussion by tkriplean** [hyperdiv/hyperdiv#55] [Help]
Maybe I'm missing something obvious, but is there a way to add an external script to a hyperdiv app? I couldn't figure out how to leverage the plugin system for it. I'm trying to link to a Plausible Analytics script.

**Comment by mariusnita** [comment]
There's no official way currently, but I'm working on adding a way to load custom JS/CSS into the index page.

This issue shows a workaround for adding a custom CSS script: https://github.com/hyperdiv/hyperdiv/issues/21. Maybe adding a `<script src="{external url path}"></script>` to the body using that technique might work for now.

You could also possibly use the current plugin API by making a plugin with a `js-link` pointing to the analytics URL, and then instantiating that plugin at the top level of the app:

```py
class Analytics(hd.Plugin):
  _assets = [
    hd.Plugin.js_link('https://link/to/analytics.js'),
    hd.Plugin.js('(() => { window.hyperdiv.registerPlugin("Analytics", () => () => {}) })();')
  ]
  
def main():
  Analytics()
  ... rest of app code ...
```

Basically a plugin that does nothing except load an external JS URL. I haven't verified if this works.

Note that these are unofficial workarounds that may/will break in future releases but hopefully the code fixes will be simple.

**Comment by tkriplean** [comment]
Thanks Marius. The js_link approach doesn't work for me because I also need to set data attributes on the script. 

I've tried doing the plugin route but have run into a variety of challenges. For example, one of the scripts I've tried to load via a plugin generates Chrome error: "Failed to execute 'write' on 'Document': It isn't possible to write into a document from an asynchronously-loaded external script unless it is explicitly opened". 

However, I did just get the analytics script to work with the plugin. I had been having trouble because when I added the script to the shadow_root, it wasn't exposed to document.querySelector, which the external script required. However, adding to document.body works.

In case it is useful to others as we await official support, this is my plugin:

plugins/script/script.py
```
import hyperdiv as hd
import os


class Script(hd.Plugin):
    defn = hd.Prop(hd.String, None)

    _assets = [
        (
            "js-link",
            os.path.join(os.path.dirname(__file__), "assets", "script.js"),
        )
    ]
```

plugins/script/assets/script.js
```
(function() {

const hd = window.hyperdiv

hd.registerPlugin('Script', function(key, shadow_root, initial_props) {

    const parser = new DOMParser()
    const doc = parser.parseFromString(initial_props.defn, 'text/html')
    const scriptFromString = doc.querySelector('script')

    const script = document.createElement('script')

    let attrs = ""
    for (let attr of scriptFromString.attributes) {
      script.setAttribute(attr.name, attr.value);
      attrs += "-" + attr.name + ":" + attr.value
    }

    let script_key = encodeURI(attrs)
    if (!document.body.querySelector("[data-script-key=\"" + script_key + "\"]")){
        script.setAttribute("data-script-key", script_key)
        document.body.appendChild(script) // note: adding to shadow_root will cause problems
    }

    return function(prop_key, prop_value) {};
});

})();
```

Usage: 

```
from plugins.script.script import Script

def my_view():

    Script(
        defn='<script defer data-domain="my.com" src="https://external.script.js"></script>'
    )
    // other app code...   

```

**Comment by mariusnita** [comment]
Wild! Sorry you have to do that workaround for now, will add support for injecting scripts soon.
